### PR TITLE
fix(UseMutationResult): result type should be more specific to allow type narrowing

### PR DESCRIPTION
**Pull request by raoufswe**
fixes https://github.com/tannerlinsley/react-query/issues/2601

This PR refactors useMutationResult type to follow the same approach/logic of useQueryResult type.

### CODE REVIEW

File: src/core/mutationObserver.ts [hunk 1]
```diff
@@ -137,7 +137,7 @@ export class MutationObserver<
       isIdle: state.status === 'idle',
       mutate: this.mutate,
       reset: this.reset,
-    }
+    } as MutationObserverResult<TData, TError, TVariables, TContext>
```
Comment (TkDodo) [inline src/core/mutationObserver.ts:140]: could we maybe do the same thing here as we do in the queryObserver, which is: declare the type as  `MutationObserverBaseResult` and then only cast it after that, see:

https://github.com/tannerlinsley/react-query/blob/f9b23fcae9c5d45e3985df4519dd8f78a9fa364e/src/core/queryObserver.ts#L544-L569

I think this would give a bit more saftey because things we set here would need to confirm to the BaseResult type, which I think they don't need with a pure type assertion

**Review by TkDodo — CHANGES_REQUESTED** [review]

File: src/react/useMutation.ts [hunk 2]
```diff
@@ -97,9 +93,9 @@ export function useMutation<
   }, [])
 
   const mutate = React.useCallback<
-    UseMutateFunction<TData, TError, TVariables, TContext>
+    MutateFunction<TData, TError, TVariables, TContext>
   >((variables, mutateOptions) => {
-    obsRef.current!.mutate(variables, mutateOptions).catch(noop)
+    return obsRef.current!.mutate(variables, mutateOptions).catch(noop)
   }, [])
```
Comment (TkDodo) [inline src/react/useMutation.ts:99]: why this change please? It's per design that `mutate` doesn't return anything...

**Review by TkDodo — CHANGES_REQUESTED** [review]

File: src/react/useMutation.ts (hunk 2 above)
Comment (raoufswe) [reply src/react/useMutation.ts:99]: MutationFunction (core/types) and UseMutateFunction (react/types) are quite identical. The only difference is that MutationFuncation returns Promise<TData>, and useMutationFunction returns void. So I assumed that MutationFunction is more accurate since it's from the core package. 

If this isn't true, I'm happy to revert the change

Comment (TkDodo) [reply src/react/useMutation.ts:99]: yes please. `.mutate` shouldn't return a promise. we've had that in v2 and it was confusing for many, which is why we added a dedicated `mutateAsync` for that case.

Comment (TkDodo) [reply src/react/useMutation.ts:99]: what's the status here please? There seem to be conflicts now with the new release we just made, please have a look at resolving them :)

Comment (raoufswe) [reply src/react/useMutation.ts]: sorry to keep this on hold, Dominik

I have reverted the .mutate change and solved the conflict, could you please have a look now?

thanks for your time.

**Review by TkDodo — APPROVED** [review]

**Comment by TkDodo** [comment]
hmm, for some reason, the type narrowing doesn't seem to work in the codesandbox preview builds. Can you have a look?

https://codesandbox.io/s/tannerlinsley-react-query-basic-typescript-forked-6zf7r?file=/src/index.tsx

File: src/react/types.ts [hunk 3]
```diff
@@ -119,24 +119,22 @@ export type UseMutateAsyncFunction<
...
-  failureCount: number
-  isError: boolean
-  isIdle: boolean
-  isLoading: boolean
-  isPaused: boolean
-  isSuccess: boolean
+> = Exclude<
+  MutationObserverResult<TData, TError, TVariables, TContext>,
+  'mutate'
+> & {
```
Comment (raoufswe) [inline src/react/types.ts:130]: @TkDodo I'm trying to override the type of mutate here and define it with UseMutateFunction so I can fix this type error

Screenshot 2021-09-08 at 11 09 12 PM

I tried to used Omit, but it broke the union in MutationObserverResult and I can't change it to interface and then extend as MutationObserverResult is a type 

do you have any idea what the best approach would be here?

**Review by TkDodo — APPROVED** [review]
